PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/UseOfBlockOnAsyncClientsCheck.java

```diff
@@ -0,0 +1,170 @@
+
+            if (qualifierExpression instanceof PsiMethodCallExpression) {
+                PsiMethodCallExpression qualifierMethodCall = (PsiMethodCallExpression) qualifierExpression;
+
+                // Get the return type of the qualifier method call
+                PsiType qualifierReturnType = qualifierMethodCall.getType();
+
+                if (qualifierReturnType instanceof PsiClassType) {
+                    PsiClass qualifierReturnTypeClass = ((PsiClassType) qualifierReturnType).resolve();
+
+                    // Check if the return type is a subclass of Mono or Flux
+                    if (qualifierReturnTypeClass != null && isReactiveType(qualifierReturnTypeClass)) {
+
+                        return isAzureAsyncClient(qualifierMethodCall);
+                    }
+                }
+            }
+            return false;
+        }
+
+        /**
+         * This method is used to check if the method call is on an async client in Azure SDK.
+         * The method call is checked to see if it is on an async client in Azure SDK.
+         *
+         * @param qualifierMethodCall PsiMethodCallExpression - the method call expression to check
+         * @return true if the method call is on an async client in Azure SDK, false otherwise
+         */
+        private boolean isAzureAsyncClient(PsiMethodCallExpression qualifierMethodCall) {
+
+            // Get the expression that calls the method returning a reactive type
+            PsiExpression clientExpression = qualifierMethodCall.getMethodExpression().getQualifierExpression();
+
+            // Travel up the method call chain to get the client expression
+            while (clientExpression instanceof PsiMethodCallExpression) {
+
+                clientExpression = ((PsiMethodCallExpression) clientExpression).getMethodExpression().getQualifierExpression();
+            }
+
+            // a ReferenceExpression is the last expression in the chain - the client object
+            if (clientExpression instanceof PsiReferenceExpression) {
+                PsiType clientType = clientExpression.getType();
+
+                if (clientType instanceof PsiClassType) {
+                    PsiClass clientClass = ((PsiClassType) clientType).resolve();
+
+                    return clientClass != null && clientClass.getQualifiedName().startsWith(RuleConfig.AZURE_PACKAGE_NAME) && clientClass.getQualifiedName().endsWith("AsyncClient");
+                }
+            }
+            return false;
+        }
+
+        /**
+         * Helper method to check if the class / return type is a subclass of Mono or Flux.
+         *
+         * @param psiClass PsiClass - the class to check
+         * @return true if the class is a subclass of Mono or Flux, false otherwise
+         */
+        private boolean isReactiveType(PsiClass psiClass) {
+            return RULE_CONFIG.getListedItemsToCheck().contains(psiClass.getQualifiedName());
```

Comment:
How are we checking the sub-types of Mono and Flux?

Reply:
Retrieve the return type of the method call.
Check if the return type is a PsiClassType.
Resolve the PsiClass from the PsiClassType.
Check if the resolved class is a sub-type of Mono or Flux by comparing its qualified name against Flux/Mono qualified names